packages/edit-site/src/store/selectors.js

```diff
@@ -247,6 +248,61 @@ export function isSaveViewOpened( state ) {
 	return state.saveViewPanel;
 }
 
+/**
+ * Helper method to iterate through all blocks, recursing into inner blocks.
+ * Returns a flattened array of all blocks.
+ *
+ * @param {Array} blocks Blocks to flatten.
+ *
+ * @return {Array} Flattened object.
+ */
+function flattenBlocks( blocks ) {
+	const result = [];
+
+	const stack = [ ...blocks ];
+	while ( stack.length ) {
+		const { innerBlocks, ...block } = stack.shift();
+		// Place inner blocks at the beginning of the stack to preserve order.
+		stack.unshift( ...innerBlocks );
+		result.push( block );
+	}
+
+	return result;
+}
+
+function getFilteredTemplatePartBlocks( templateBlocks = [], templateParts ) {
+	const templatePartsById = templateParts
+		? // Key template parts by their ID.
+		  templateParts.reduce(
+				( newTemplateParts, part ) => ( {
+					...newTemplateParts,
+					[ part.id ]: part,
+				} ),
+				{}
+		  )
+		: {};
+
+	return flattenBlocks( templateBlocks )
+		.filter( ( block ) => isTemplatePart( block ) )
+		.map( ( block ) => {
```

Comment:
Here we're looping four times: inside flatter, the filter, the map and another filter. If we want maximum performance, we could copy the logic of "flattenBlocks" here and add the checks and mapping within the "while" loop. 

That said, the improvement could be too small to worth it in terms of code clarity, so really your call here, just wanted to comment it on that.

Reply:
We could as well combine the additional filtering and mapping with `flattenBlocks()`, so it all would happen in that `while` block - `continue`-ing if the block is not a template block; mapping and again `continue`-ing if there's no `templatePart`. That would just be a single loop instead of several ones.